### ISSUE TITLE
RelatedObjectDoesNotExist problem

### ISSUE THREAD

**Issue by bodgerbarnett** [mcclory/django-dynamic-storages#3]
I'm interested in this technique for separating my media into different directories for each of the teams my site has.

But I can't seem to get it to work.

For an example model that has a reference to Team and a file, I want to use the function mentioned in the docs to get the storage:-

```
def get_storage(instance):
    if instance and getattr(instance, "team", None):
        tst = TeamStorageTarget.for_team(instance.team)
        if tst:
            return tst.storage_backend
    return default_storage
```

 But when the file is first created and saved, the object hasn't got a team yet - so it uses the default storage. Whereas I want it to use my team storage.

What am I doing wrong here? I wonder whether my code should create the object first - and then save the file? Then the object *would* have a team and it would use the right storage. Is that the correct way to use it?

**Comment by longhotsummer** [comment]
> What am I doing wrong here? I wonder whether my code should create the object first - and then save the file? Then the object _would_ have a team and it would use the right storage. Is that the correct way to use it?

Yes, that would be my approach, or at least set `instance.team` without saving. Even if you weren't using this library, the code that decides where to put the file has to have sufficient context to do so.

**Comment by mcclory** [comment]
So... full disclosure, this was a part of a project that kinda died on me and I'm just now getting back to looking at really keeping up with this.  

There is a problem of sequencing here with the process as @longhotsummer mentioned - the underlying objects are assumed to exist/have been created.  This mostly follows the same pattern as the callable for `upload_to` within the `django.db.models.fields.FileField` object as documented here: https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.FileField.upload_to

It'd be kinda cool if django did this by default, but the default implementation of FileField does take `storage` as a callable, but that callable doesn't take any inputs: https://github.com/django/django/blob/097e3a70c1481ee7b042b2edd91b2be86fb7b5b6/django/db/models/fields/files.py#L243-L252

I'm going to be playing around with this a bit more and would love any input you guys might have.
